presto-benchmark-driver/src/main/java/com/facebook/presto/benchmark/driver/BenchmarkQueryRunner.java

```diff
@@ -149,53 +152,56 @@ public BenchmarkQueryResult execute(Suite suite, ClientSession session, Benchmar
-
-            // read query output
             ImmutableList.Builder<String> schemas = ImmutableList.builder();
-            while (client.isValid() && client.advance()) {
-                // we do not process the output
-                Iterable<List<Object>> data = client.currentData().getData();
-                if (data != null) {
-                    for (List<Object> objects : data) {
-                        schemas.add(objects.get(0).toString());
-                    }
-                }
-            }
-
-            // verify final state
-            if (client.isClosed()) {
-                throw new IllegalStateException("Query aborted by user");
-            }
-
-            if (client.isGone()) {
-                throw new IllegalStateException("Query is gone (server restarted?)");
+            AtomicBoolean success = new AtomicBoolean(true);
+            execute(
+                    session,
+                    "show schemas",
+                    queryData -> {
+                        for (List<Object> objects : queryData.getData()) {
+                            schemas.add(objects.get(0).toString());
+                        }
+                    },
+                    queryError -> {
+                        success.set(false);
+                        handleFailure(getCause(queryError));
+                    });
+            if (success.get()) {
+                return schemas.build();
             }
+        }
+    }
 
-            QueryError resultsError = client.finalStatusInfo().getError();
-            if (resultsError != null) {
-                RuntimeException cause = null;
-                if (resultsError.getFailureInfo() != null) {
-                    cause = resultsError.getFailureInfo().toException();
-                }
-                handleFailure(cause);
-
-                continue;
-            }
+    private StatementStats execute(ClientSession session, String name, String query)
+    {
+        return execute(
+                session,
+                query,
+                queryData -> {}, // we do not process the output
+                resultsError -> {
+                    throw new BenchmarkDriverExecutionException(format("Query %s failed: %s", name, resultsError.getMessage()), getCause(resultsError));
+                });
+    }
 
-            return schemas.build();
+    private static RuntimeException getCause(QueryError queryError)
+    {
+        if (queryError.getFailureInfo() != null) {
+            return queryError.getFailureInfo().toException();
         }
+        return null;
     }
 
-    private StatementStats execute(ClientSession session, String name, String query)
+    private StatementStats execute(ClientSession session, String query, Consumer<QueryData> queryDataConsumer, Consumer<QueryError> queryErrorConsumer)
     {
         // start query
         StatementClient client = new StatementClient(okHttpClient, session, query);
```

Comment:
`StatementClient` needs to be closed. I'm fixing it as part of my big PR: https://github.com/prestodb/presto/pull/9607 (commit number 2). However, since you're making changes to this class you can fix it as well.